src/merge.rs

```diff
@@ -1796,62 +1788,98 @@ impl<'t> MergedRoot<'t> {
     pub fn completion_ops(&self) -> CompletionOps<'_> {
         let mut ops = CompletionOps::default();
         accumulate(&mut ops, self.node(), 1, false);
-        for guid in self.delete_locally.iter() {
-            if self.local_tree.mentions(guid) && self.remote_tree.mentions(guid) {
-                // Only flag tombstones for items that exist in both trees,
-                // excluding ones for items that don't exist in the local
-                // tree. This can be removed once we persist tombstones in
-                // bug 1343103.
-                let flag_as_merged = FlagAsMerged(guid);
-                ops.flag_as_merged.push(flag_as_merged);
+        for guid in self.deletions() {
+            match (
+                self.local_tree.record_for_guid(guid),
+                self.remote_tree.record_for_guid(guid),
+            ) {
+                (Record::Unmentioned, Record::Unmentioned) => {
+                    // The item isn't mentioned (existing or deleted) on either
+                    // side, so nothing to do.
+                }
+                (Record::Unmentioned, Record::Exists(remote_node)) => {
+                    // The item isn't mentioned locally, but exists remotely.
+                    // This is the case for invalid remote-only items. Upload a
+                    // tombstone, but don't flag it as remotely merged; we'll do
+                    // that when we store the uploaded tombstone.
+                    ops.insert_local_tombstones
+                        .push(InsertLocalTombstone(remote_node));
+                    ops.upload_tombstones
+                        .push(UploadTombstone(&remote_node.item().guid));
+                }
+                (Record::Unmentioned, Record::Deleted) => {
+                    // The item isn't mentioned locally, but is deleted
+                    // remotely. This is the case for a first sync, and will
+                    // change if we store tombstones (bug 1343103). For now,
```

Comment:
This does mean we'll need to merge all tombstones again if we ever decide to store them, but we can cross that bridge later. Marking remote tombstones as merged also means we can filter them out in subsequent merges, which is going to make application more efficient.